omen2/object.py

```diff
@@ -0,0 +1,252 @@
+                        "will not create %s without primary key"
+                        % self.__class__.__name__
+                    )
+                assert not need_id_field
+                need_id_field = field
+        return need_id_field
+
+    def __setattr__(self, k, v):
+        if k[0] == "_":
+            super().__setattr__(k, v)
+            return
+
+        if self._meta and self._meta.table and not self._meta.locked:
+            raise OmenError("use with: protocol for bound objects")
+
+        if self._meta and not hasattr(self, k):
+            raise AttributeError("Attribute %s not defined" % k)
+        super().__setattr__(k, v)
+
+    def _get_related(self):
+        related = {}
+        for val in self.__dict__.values():
+            if isinstance(val, Relation) and val.cascade:
+                related[val] = list(val.select())
+        return related
+
+    def _collect_cascade(self):
+        # cascading id changes to relations is expensive, and involves weird swaps
+        # but it seems like this should be rare behavior
+        # we save a list of all related objects here
+        # and then later, we go throught and update all of them
+        # because we don't know what the lambda-relationship is
+        if self._meta.pk:
+            pk = self._to_pk()
+            if pk != self._meta.pk:
+                for k, v in self._meta.pk.items():
+                    setattr(self, k, v)
+                try:
+                    return self._get_related()
+                finally:
+                    for k, v in pk.items():
+                        setattr(self, k, v)
+        return {}
+
+    def _commit(self):
+        cascade = self._collect_cascade() if self._cascade else {}
+
+        if self._meta.table:
+            self._save()
+
+        for val in self.__dict__.values():
+            if isinstance(val, Relation):
+                val.commit(self._meta.table.manager)
+
+        for rel, objs in cascade.items():
+            for obj in objs:
+                rel._link_obj(obj)
+
+    def _remove(self):
+        cascade = self._get_related() if self._cascade else {}
+
+        if self._meta.table:
+            table = self._meta.table
+            table._remove(self)
+
+        for rel, objs in cascade.items():
+            for obj in objs:
+                rel.remove(obj)
+
+    def _save(self):
+        need_id_field = self._need_id()
+        table = self._meta.table
+        if need_id_field or self._meta.new:
+            table.insert(self, need_id_field)
+        else:
+            table.update(self)
+
+        self._meta.new = False
+        self._save_pk()
+
+    def _rollback(self):
+        pk = self._to_pk()
+        db_row = self._meta.table.db_select(pk)
+        if db_row:
+            if len(db_row) > 1:
+                raise OmenMoreThanOneError("more than 1 %s in the db" % self.__class__)
+        self._update(db_row[0])
+
+    def __enter__(self):
```

Comment:
Doesn't this need to return a copy? Else any other thread with a reference to the object will see my modifications (and also be able to modify, themselves)

Reply:
resolved in track-changes MR